syft/execution/protocol.py

```diff
@@ -32,12 +33,16 @@ class func2protocol(object):
     This class should be used only as a decorator.
     """
 
-    def __init__(self, args_shape=None):
+    def __init__(self, roles: list = [], args_shape: dict = {}):
         self.args_shape = args_shape
+        self.role_names = roles
 
     def __call__(self, protocol_function):
         # create the roles present in decorator
-        roles = {role_id: Role() for role_id in self.args_shape.keys()}
+        roles = {
+            role_id: Role(worker=VirtualWorker(id=role_id, hook=sy.local_worker.hook))
```

Comment:
Isn't `sy.local_worker.hook` the default, here?

Reply:
Oh, maybe so!

Reply:
Hmm, I don't think it is, but maybe it should be?

Reply:
Oh, sorry, I thought it was!
It could, or we could also have a "local_hook" instead (or with) the "local_worker", not sure I see everything but I find this `sy.local_worker.hook` a bit odd

Reply:
Completely agree on it being odd! Workers and hooks are coupled in some weird ways we'd probably have to untangle.